drivers/sensor/hx711/hx711.c

```diff
@@ -0,0 +1,622 @@
+
+#define DT_DRV_COMPAT avia_hx711
+
+#include <device.h>
+#include <drivers/gpio.h>
+#include <drivers/sensor.h>
+#include <drivers/sensor/hx711.h>
+#include <irq.h>
+#include <logging/log.h>
+#include <pm/device.h>
+#include <pm/device_runtime.h>
+#include <string.h>
+#include <sys/byteorder.h>
+#include <zephyr.h>
+
+struct hx711_data {
+	struct gpio_callback dout_gpio_cb;
+	struct k_sem dout_sem;
+
+	const struct hx711_config *cfg;
+
+	int32_t reading;
+
+	int offset;
+	struct sensor_value slope;
+	struct sensor_value calibration_weight;
+	int sample_fetch_timeout;
+	char gain;
+	enum hx711_rate rate;
+	const int32_t cw[2];
+	const int32_t sl[2];
+};
+
+struct hx711_config {
+	int instance;
+
+	struct gpio_dt_spec dout;
+	struct gpio_dt_spec sck;
+	struct gpio_dt_spec rate;
+};
+
+LOG_MODULE_REGISTER(HX711, CONFIG_SENSOR_LOG_LEVEL);
+
+static void hx711_gpio_callback(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
+{
+	struct hx711_data *data = CONTAINER_OF(cb, struct hx711_data, dout_gpio_cb);
+	const struct hx711_config *cfg = data->cfg;
+
+	gpio_pin_interrupt_configure_dt(&cfg->dout, GPIO_INT_DISABLE);
+
+	/* Signal thread that data is now ready */
+	k_sem_give(&data->dout_sem);
+}
+
+/**
+ * @brief Send a pulse on the SCK pin.
+ *
+ * @param data Pointer to the hx711 driver data structure
+ *
+ * @retval The value of the DOUT pin (HIGH or LOW)
+ *
+ */
+static int hx711_cycle(const struct device *dev)
+{
+	const struct hx711_config *cfg = dev->config;
+
+	/* SCK set HIGH */
+	gpio_pin_set_dt(&cfg->sck, true);
+	k_busy_wait(1);
+
+	/* SCK set LOW */
+	gpio_pin_set_dt(&cfg->sck, false);
+	k_busy_wait(1);
+
+	/* Return DOUT pin state */
+	return gpio_pin_get_dt(&cfg->dout);
+}
+
+/**
+ * @brief Read HX711 data. Also sets GAIN for the next cycle.
+ *
+ * @param dev Pointer to the hx711 device structure
+ * @param chan Channel to fetch data for.
+ *             Only SENSOR_CHANNEL_WEIGHT is available.
+ *
+ * @retval 0 on success,
+ * @retval -EACCES error if module is not powered up.
+ * @retval -EIO error if sample_fetch_timeout msec's elapsed with no data available.
+ *
+ */
+static int hx711_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+
+	int ret = 0;
+	uint32_t count = 0;
+	int i;
+
+	struct hx711_data *data = dev->data;
+	const struct hx711_config *cfg = dev->config;
+
+	ret = pm_device_runtime_get(dev);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (gpio_pin_get_dt(&cfg->dout) == 1) {
+		k_sem_take(&data->dout_sem, K_NO_WAIT);
+
+		if (gpio_pin_interrupt_configure_dt(&cfg->dout,
+						    GPIO_INT_EDGE_TO_INACTIVE) < 0) {
+			LOG_ERR("Failed to set dout GPIO interrupt");
+		}
+
+		/* Wait until reading is ready */
+		if (k_sem_take(&data->dout_sem, K_MSEC(data->sample_fetch_timeout))) {
+			LOG_ERR("Weight data not ready within %d ms. "
+				"Is the device properly connected?",
+				data->sample_fetch_timeout);
+			return -EIO;
+		}
+	}
+
+	/* Clock data out. Optionally disable interrupts */
```

Comment:
For the rare case that some application finds measuring weight more important than serving other routines.
On second thought this might just better be removed.